### ISSUE TITLE
P0 score: validate PM2S as a production score-MIDI challenger

### ISSUE THREAD

**Issue by gr-rr** [ListenCloser/listencloser#953]
Parent: #945. Representation-quality owner: #498. Active implementation: PR #955.

## Product objective

Treat **PM2S as a real, production-shaped candidate** for the `solo_piano` Score path:

```text
TransKun
  -> canonical performance MIDI
  -> PM2S
  -> score MIDI
  -> MuseScore MIDI import
  -> MusicXML
  -> OSMD
```

PM2S is not deprecated by older research or by another project's decision to omit parts of it. That external evidence is useful context, but it is **not a repository-level rejection gate** for ListenCloser. Our SOT is current product evidence on our own pipeline.

## Why PM2S is viable now

- source is MIT-licensed;
- pretrained inference artifacts are publicly available and can be pinned/provenanced;
- it directly models the missing `performance MIDI -> score MIDI` stage;
- it fits the existing notation-engine seam without replacing canonical TransKun performance evidence;
- it is deployable enough to test in the actual product now, unlike stronger-looking candidates whose current packaging/model terms are not yet production-usable.

This makes PM2S a **first-class experimental challenger**, not a reference-only technology and not a pre-judged winner.

## Semantic contract

- TransKun performance MIDI remains canonical for Piano Roll/performance evidence.
- PM2S output is a distinct derived **score-MIDI** artifact with exact source/model provenance.
- PM2S must never overwrite or become the canonical performance MIDI.
- MuseScore MIDI import is a downstream notation stage and may make additional notation decisions; provenance must make that explicit.
- Selecting PM2S must be fail-closed: PM2S failure cannot silently feed the original performance MIDI through the MuseScore-direct path and label the result PM2S.
- During comparison, MuseScore-direct remains the control/baseline, not a permanent hidden fallback.

## Current implementation

PR #955 is the active implementation. It supersedes closed #946 with corrected stage semantics and keeps PM2S score MIDI distinct from performance MIDI.

Near-term completion requires:

- [ ] land the bounded PM2S backend/runtime adapter with pinned source/model identity;
- [ ] expose explicit engine selection for product/developer comparison;
- [ ] run real production-shaped `solo_piano` inference through PM2S;
- [ ] verify valid score-MIDI -> MusicXML -> OSMD render/playback/seek;
- [ ] compare the same real Works against MuseScore-direct;
- [ ] inspect catastrophic note deletion/hallucination, rhythm/measure validity, tie/rest fragmentation, staff/voice organization, and readability;
- [ ] preserve exact PM2S provenance and untouched TransKun Piano Roll evidence;
- [ ] record checkpoint usage/licensing boundary before any commercial distribution assumption.

## Decision rule

Do **not** reject PM2S because of historical papers, prior experiments, or another repository's architecture choice. Likewise, do not promote it merely because it runs.

After it is runnable in ListenCloser:

- promote it for the validated piano Score route if current product/reference evidence shows a material improvement and operational/legal gates are acceptable;
- otherwise record the concrete failure and move to the next viable learned candidate;
- once a winner is established, remove superseded interpretation paths rather than maintaining a permanent fallback tree.

## Non-goals

- no generic benchmark framework as a prerequisite;
- no handcrafted quantization stack layered around PM2S to force a win;
- no conflation of score MIDI with performance MIDI;
- no claim that MuseScore is a semantically lossless serializer;
- no permanent multi-engine fallback tree.

Refs #945 #955 #498 #613 #948 #946.

**Comment by gr-rr** [comment]
Implementation started in #955.

#955 is the clean current-main successor to closed #946. It implements the backend/runtime half of this issue with the corrected semantic boundary:

```text
performance MIDI -> PM2S -> score MIDI -> MuseScore MIDI import -> MusicXML
```

The PM2S score MIDI is preserved as the returned notation MIDI; MuseScore receives only that derived MIDI; MuseScore is explicitly recorded as a downstream MIDI-import stage rather than a passive serializer; PM2S failures propagate without silently switching to the direct MuseScore path.

The first protected Build is queued. This issue remains open for the Score-engine UI selector, production-shaped real inference proof, and real-work A/B before canonization.

**Comment by gr-rr** [comment]
Production-shaped same-performance A/B is complete (run 33472827558, artifact `score-ab-real-piano-1`). This is deliberately a smoke comparison, not a canonization benchmark.

Fixture / shared evidence:
- `tests/fixtures/real-piano.m4a`
- TransKun 2.0.1 run once: 102 performance-MIDI notes, 21.39s CPU
- Beat This 1.1.0 run once: 57 beats / 31 downbeats, ~75 BPM
- Both Score engines consumed the same performance MIDI / beat evidence

MuseScore direct:
- score stage: 0.447s
- notation MIDI: 102 notes; pitch precision/recall vs performance MIDI 1.000 / 1.000
- MusicXML: 36 measures, inferred 3/4, 173 pitched notes, 114 rests, 142 ties, 20 beams, 0 tuplets, 5 unique duration values

PM2S -> MuseScore:
- score stage: 0.946s
- score MIDI: 101 notes; pitch precision/recall vs performance MIDI 1.000 / 0.990
- MusicXML: 14 measures, inferred 4/4, 154 pitched notes, 145 rests, 104 ties, 41 beams, 43 tuplets, 15 unique duration values

Interpretation:
- PM2S is materially changing score semantics before MuseScore; this is not merely a cosmetic re-export. Meter/measure/rhythm structure diverges substantially.
- On this one fixture PM2S is ~2.1x slower at the Score stage, though both are sub-second and transcription dominates wall-clock latency.
- Note preservation is effectively intact (one note lost by PM2S), but structural counts alone do **not** tell us which score is more readable/correct. In particular, 43 tuplets + 15 duration values is a reason to inspect PM2S visually rather than assuming the learned route wins.
- Keep MuseScore as default and PM2S experimental for now. The result supports exposing both and comparing them, not canonizing either from one piece.

Raw `performance.mid`, both derived MIDIs, both MusicXML files, and `summary.json` are retained in the workflow artifact.

**Comment by gr-rr** [comment]
Started the next bounded validation slice in #1001. The existing production real-stack golden path only exercises the default MuseScore route, so #1001 keeps that control untouched and adds a second isolated real-stack journey that selects **Solo piano + PM2S** through the actual Processing UI, imports the rights-safe real-piano fixture, requires canonical Piano Roll notes, renders/plays Score, and verifies succeeded durable job parameters plus persisted PM2S Version provenance. This is validation only: no routing/promotion decision or fallback change. The protected lane’s wall-clock behavior will also tell us whether the production-shaped TransKun + PM2S path is operationally viable.

**Comment by gr-rr** [comment]
## OSS scout implication for the current PM2S decision

The broader MIR/score survey does **not** justify replacing this issue's current product comparison with a different engine-shopping exercise.

The same-performance A/B already established the useful facts that external papers cannot answer for us:

- both routes consume the same canonical TransKun performance evidence;
- PM2S preserves pitch content almost completely but materially rewrites meter/rhythm/measure structure;
- PM2S is still sub-second at the Score stage on the fixture, so operational latency is not the present blocker;
- its 43 tuplets / 15 duration values on that one Work are a concrete readability warning, while structural counts alone cannot prove MuseScore is correct either.

Current external challenger status reinforces keeping this comparison focused:

- Rubato is a higher reported-quality direct-score target but is not presently a verified deployable/open production dependency;
- MIDI2ScoreTransformer is a stronger modular literature reference but still has licensing/custom-dependency barriers;
- joint-apt-epr is permissively coded but research-shaped and should only become the next bounded feasibility candidate if PM2S is rejected.

Therefore the next decision-changing evidence remains **multi-piece current-product readability/fidelity validation of PM2S vs MuseScore-direct**, with exact provenance and untouched Piano Roll authority. Do not add another Score engine before that decision, and do not patch PM2S's tuplet/rhythm behavior with bespoke heuristics to manufacture a win.

If representative product evidence confirms the current warning tail, record a concrete `REJECT` and return to #945's next production-eligible candidate. If PM2S materially wins readability despite the structural differences, promote only for the validated domain and remove the superseded route rather than creating a permanent fallback tree.

**Comment by gr-rr** [comment]
MIR research readout after re-checking #1056/#953 ownership and current evidence:

**Highest-value next evidence: a small, rights-safe multi-piece Score A/B focused on human-readable score correctness, not another infrastructure benchmark.**

#1056 is already a negative/research result: Omnizart Stream is weakly matched to the dense-mixture target, Pop lacks sufficient training-data/checkpoint provenance, and checkpoint commercial-use terms are unresolved. Further Omnizart inference would not currently change the production decision. [#1056 preflight]

For PM2S, the one-piece `real-piano.m4a` smoke A/B is already informative but non-decisive: PM2S preserved 101/102 performance notes and ran ~2.1x the Score-stage time, while changing 36 measures/3-4 meter/5 durations (MuseScore-direct) to 14 measures/4-4 meter/15 durations/43 tuplets (PM2S->MuseScore). Those structural differences make **readability/correctness** the unresolved question, not basic runtime viability.

Recommended next evidence slice:

1. Use 3–5 representative, rights-safe solo-piano passages spanning simple melody, chordal accompaniment, sustained/pedal material, and rhythmically complex material.
2. Transcribe once; feed identical canonical performance MIDI + beat evidence to MuseScore-direct and PM2S.
3. Produce paired MusicXML/OSMD renders and a compact per-piece rubric: note deletion/hallucination, meter/measure validity, rhythmic quantization, tie/rest fragmentation, voice/staff organization, tuplets, and overall readability.
4. Where aligned reference scores exist, add objective note/rhythm metrics; otherwise treat expert visual/listening inspection as primary evidence rather than inventing a proxy score.
5. Record score-stage latency, failures, and exact provenance/checkpoint identity; keep Piano Roll/performance MIDI untouched.

**Decision threshold:** promote only if PM2S wins materially on readability/fidelity across the small corpus without unacceptable regressions. If it loses consistently, reject/remove the experimental path rather than layering repair heuristics. If it is clearly better but operational evidence remains blocked by the failing protected real-stack lane, classify as promising/research and fix that gate separately.

I would not spend the next MIR slice on a generic benchmark framework or more Omnizart work. The marginal information gain is highest from the PM2S multi-piece human-readable A/B.

**Comment by gr-rr** [comment]
### 2026-09-03 provenance/licensing gate: PM2S is not promotable yet

Fresh verification found a concrete reproducibility problem in the current PM2S runtime identity:

- ListenCloser records the PM2S source as `JonghoKimSNU/PM2S` with revision `main` and checkpoint reference `zenodo:10520196:scraper-v3-and-diffusion-500k` (`backend/engines/notation/pm2s_engine.py`, `backend/Dockerfile`).
- The current GitHub repository API returns **404** for `JonghoKimSNU/PM2S`.
- A different live repository named `cheriell/PM2S` exists and is MIT-licensed, but it is **not safe to silently substitute**: its implementation/checkpoint layout is the 2022 neural beat-tracking PM2S (`RNNJointBeatModel.pth`, etc.), whereas ListenCloser's runtime expects Scraper + diffusion checkpoints (`env-Scraper_epoch-003_val_loss-0.00205757.pt`, `diffusion-step=500000.pt`).
- `main` is also a mutable source revision, not an immutable production identity.
- Production-image PM2S smoke is still green on amd64/arm64, so runtime viability remains proven; this finding is specifically about reproducibility/provenance and the commercial-distribution gate, not whether the current image can execute.
- I have not found authoritative license metadata for the exact Zenodo checkpoint pair yet. Treat checkpoint redistribution/commercial usage as **unresolved**, not implicitly covered by a similarly named repository's MIT source license.

Decision impact: **KEEP EXPERIMENTAL** at minimum until the exact source repository/commit and checkpoint license are independently resolvable and pinned. Do not promote PM2S based on the existing single-piece A/B evidence.

I also recovered the original bounded A/B workflow from run `33472827558` (`.github/workflows/score-ab-real-piano.yml` at `6521332a`). It transcribes once, then sends the exact same canonical performance MIDI + beat evidence through MuseScore and PM2S. I will reuse that shape for the remaining multi-piece readability/correctness evidence rather than building a benchmark framework.

**Comment by gr-rr** [comment]
## PM2S NEXT_PROBE decision — retain current production route

**Authority:** `c8d830c650cdc3e6f4c35f63d369bf8923d7be47`  
**Mode:** evaluation only — no routing change, selector change, new engine, or PR.

### Ownership / prior-evidence inventory

Before running anything, I checked the existing `eval/953-*` / `score/953-*` branches and open PRs:

- `eval/953-pm2s-maestro-readability` is historical/stale relative to current authority; it only staged a push-gated probe workflow and had no trigger marker / active run.
- `score/953-pm2s-realstack-proof` is historical evidence associated with already-landed work.
- No open PR or active #953 evaluation owner was present, so I ran one **sequential** bounded probe rather than starting parallel ownership.

Probe branch: `eval/953-pm2s-next-probe-c8d830`  
Successful run: `33933568055`  
Evidence artifact: `issue-953-pm2s-next-probe-1` (artifact id `9959461845`, digest `sha256:ce3c8ec6ee27f24384a094c06ad81849bd0363d0e1f93e558d3f4a935f3e06c7`)

### Method

Three 10-second excerpts from real MAESTRO test performances were sliced from existing dataset material. Each engine received the **same performance MIDI**; source-MIDI beat/downbeat evidence was supplied to the production API call. The exact authority backend image was built before measurement.

Important honesty note: current provenance reports `beat_grid_consumed=false` for both routes here, so this probe is a same-performance-MIDI A/B, not evidence that either engine honored the supplied beat grid.

Measured outputs included score-MIDI note/pitch retention, timing deltas, runtime, and MusicXML structural counts. A second MuseScore CLI MusicXML→PNG render attempt failed with exit 40 for **both** engines, so there is no honest visual-render winner from this run. Structural XML cleanliness is reported only as a readability proxy.

### Machine-readable decision

```json
{
  "schema": "listencloser.eval.953.decision.v1",
  "authority_sha": "c8d830c650cdc3e6f4c35f63d369bf8923d7be47",
  "posture": "NEXT_PROBE",
  "decision": "RETAIN_CURRENT_ROUTE",
  "pm2s_challenger_status": "REJECT_FOR_PRODUCTION_ROUTING_AT_THIS_AUTHORITY",
  "production_change": false,
  "selector_change": false,
  "new_engine": false,
  "open_pr": false,
  "probe_branch": "eval/953-pm2s-next-probe-c8d830",
  "workflow_run_id": 33933568055,
  "artifact": {
    "id": 9959461845,
    "name": "issue-953-pm2s-next-probe-1",
    "sha256": "ce3c8ec6ee27f24384a094c06ad81849bd0363d0e1f93e558d3f4a935f3e06c7"
  },
  "corpus": {
    "dataset": "MAESTRO v3 test MIDI",
    "clips": 3,
    "seconds_per_clip": 10,
    "input_notes": 390
  },
  "metrics": {
    "musescore": {
      "successful_conversions": 3,
      "score_midi_notes": 390,
      "pitch_multiset_overlap": 390,
      "pitch_recall": 1.0,
      "score_seconds_total": 1.814837,
      "score_seconds_median": 0.602679,
      "musicxml_rests": 280,
      "musicxml_ties": 220,
      "musicxml_tuplets": 195,
      "musicxml_beams": 866,
      "png_render_successes": 0
    },
    "pm2s": {
      "successful_conversions": 3,
      "score_midi_notes": 365,
      "pitch_multiset_overlap": 365,
      "pitch_recall": 0.935897,
      "dropped_input_notes": 25,
      "per_clip_pitch_recall": [0.974194, 0.919708, 0.897959],
      "score_seconds_total": 5.643863,
      "score_seconds_median": 1.538544,
      "runtime_ratio_vs_musescore": 3.109846,
      "musicxml_rests": 224,
      "musicxml_ties": 60,
      "musicxml_tuplets": 195,
      "musicxml_beams": 424,
      "png_render_successes": 0
    }
  },
  "measurement_limits": {
    "beat_grid_consumed": false,
    "visual_png_comparison_available": false,
    "visual_render_failure": "MuseScore CLI exit 40 for both routes",
    "structural_readability_proxy_only": true
  },
  "rights": {
    "pm2s_source": "MIT",
    "pm2s_checkpoint_record": "zenodo:10520196",
    "checkpoint_production_rights": "UNRESOLVED_PER_EXISTING_ISSUE_RECORD"
  }
}
```

### Decision

**Retain the current MuseScore production route. Do not promote PM2S or add a selector.**

PM2S did produce non-empty, structurally richer MusicXML on all three real excerpts and had fewer ties/rests/beams than the current route, so the historical “nearly empty MusicXML” failure is no longer the deciding problem. However, the production-correctness bar is not met: PM2S dropped **25 / 390 input notes (6.41%)** across all three excerpts, while the current route preserved **390 / 390**. The losses were not isolated to one pathological clip (recall `97.42%`, `91.97%`, `89.80%`). PM2S was also ~`3.11x` slower in this bounded notation-stage run.

The structural-cleanliness signal is therefore research evidence, not a reason to trade away note correctness. Because correctness already fails materially, the symmetric PNG-render failure does not leave the production-routing decision ambiguous; it only means this run cannot claim a visual-readability winner. The unresolved checkpoint production-rights record would also still need resolution before any future promotion even if a later technical probe clears the correctness bar.

**#953 NEXT_PROBE result:** production challenger **not validated** at `c8d830c650cd`; current route remains authority.
